### PR TITLE
Functional testing for glance external S3 backend

### PR DESCRIPTION
**Pull request by nobuto-m**
As a prep for taking over the glance charm patch set in https://review.opendev.org/c/openstack/charm-glance/+/781748 and override #532 eventually.

Set up Ceph-Radosgw in the same model as an external S3 backend for
Glance:
- create an S3 user in Radosgw
- create an S3 bucket for Glance
- inject those parameters into Glance charm configurations as an
  external S3 backend
- upload an image to the S3 backend via Glance, and check the size of
  the image in the S3 bucket directly

### CODE REVIEW

**Comment by AurelienLourot** [comment]
Thanks for the patch!

> As a prep for taking over the glance charm patch set in https://review.opendev.org/c/openstack/charm-glance/+/781748 and override #532 eventually.

Can you please clarify: is it fully superseding #532 or not yet? In other words: can #532 then be abandoned? Thanks!

**Comment by nobuto-m** [comment]
> Thanks for the patch!
> 
> > As a prep for taking over the glance charm patch set in https://review.opendev.org/c/openstack/charm-glance/+/781748 and override #532 eventually.
> 
> Can you please clarify: is it fully superseding #532 or not yet? In other words: can #532 then be abandoned? Thanks!

This PR is ready to supercede it, and #532 can be dropped.

**Comment by AurelienLourot** [comment]
@coreycb would you like to check here if your comments from #532 have been addressed?

**Comment by AurelienLourot** [comment]
@LCVcode since this supersedes your PR #532 and thus you have expertise in this area, what do you think of this PR? Do you feel like it covers everything?

File: zaza/openstack/charm_tests/glance/setup.py [hunk 1]
```diff
@@ -161,3 +165,58 @@ def add_lts_image(glance_client=None, image_name=None, release=None,
...
+    logging.info("Getting S3 endpoint URL of Radosgw from Keystone")
+    keystone_auth = openstack_utils.get_overcloud_auth()
+    keystone_client = openstack_utils.get_keystone_client(keystone_auth)
+    endpoint_url = keystone_client.session.get_endpoint(
+        service_type="s3",
+        interface="public",
+        region="RegionOne",
+    )
+
+    logging.info("Creating a test S3 bucketfor Glance")
```
Comment (coreycb) [inline zaza/openstack/charm_tests/glance/setup.py:191]: Nit typo "bucketfor"

File: zaza/openstack/charm_tests/glance/setup.py [hunk 2]
```diff
@@ -161,3 +165,58 @@ def add_lts_image(glance_client=None, image_name=None, release=None,
...
+    bucket_name = "zaza-glance-s3-test"
+    s3_client = boto3.client(
+        "s3",
+        endpoint_url=endpoint_url,
+        aws_access_key_id=access_key,
+        aws_secret_access_key=secret_key,
+    )
+    s3_client.create_bucket(Bucket=bucket_name)
+
+    logging.info("Updating Glance configs with S3 endpoint infomation")
```
Comment (coreycb) [inline zaza/openstack/charm_tests/glance/setup.py:201]: Nit typo "infomation"

**Review by coreycb — APPROVED** [review]
Looks good to me, just a few typos.

**Review by AurelienLourot — APPROVED** [review]

**Pull request merged**